File: src/engine/pronouns.rs
use deadpool_sqlite::Pool;
use serde_json::{json, Value};

const CASES: [&str; 7] = ["prathama", "dvitiya", "tritiya", "caturthi", "panchami", "sasthi", "saptami"];

pub async fn analyze_pronoun(word: &str, pool: &Pool) -> Vec<Value> {
    let word_clone = word.to_string();
    pool.get().await.unwrap().interact(move |conn| {
        let mut results = Vec::new();
        // Look up all pronouns (we assume Pronouns were added to the DB alongside Irregulars and Numerals)
        // Note: For now we fall back to a hardcoded mapping since Pronouns didn't explicitly get their own DB table in the Python step yet.
        // We will hardcode the most common ones to match the Python script.
        let pronoun_map = vec![
            ("tad", "masculine", "prathama", vec!["saH", "tO", "te"]),
            ("tad", "masculine", "dvitiya", vec!["tam", "tO", "tAn"]),
            ("tad", "masculine", "caturthi", vec!["tasmE", "tAByAm", "teByaH"]),
            ("kim", "masculine", "prathama", vec!["kaH", "kO", "ke"]),
            ("asmad", "any", "prathama", vec!["aham", "AvAm", "vayam"]),
            ("yuzmad", "any", "prathama", vec!["tvam", "yuvAm", "yUyam"]),
        ];

        for (base, gender, case, forms) in pronoun_map {
            for (i, form) in forms.iter().enumerate() {
                if form == &word_clone {
                    let vacana = ["eka", "dvi", "bahu"][i];
                    results.push(json!({
                        "type": "pronoun",
                        "base_form": base,
                        "gender": gender,
                        "case": case,
                        "vacana": vacana
                    }));
                }
            }
        }
        results
    }).await.unwrap()
}

File: engine/pronouns.go
package engine

import "database/sql"

type pronounEntry struct {
	base   string
	gender string
	cCase  string
	forms  [3]string
}

var pronounMap = []pronounEntry{
	{"tad", "masculine", "prathama", [3]string{"saH", "tO", "te"}},
	{"tad", "masculine", "dvitiya", [3]string{"tam", "tO", "tAn"}},
	{"tad", "masculine", "caturthi", [3]string{"tasmE", "tAByAm", "teByaH"}},
	{"kim", "masculine", "prathama", [3]string{"kaH", "kO", "ke"}},
	{"asmad", "any", "prathama", [3]string{"aham", "AvAm", "vayam"}},
	{"yuzmad", "any", "prathama", [3]string{"tvam", "yuvAm", "yUyam"}},
}

// AnalyzePronoun analyzes pronouns against the hardcoded pronoun map
func AnalyzePronoun(db *sql.DB, word string) []map[string]any {
	var results []map[string]any
	vacanaList := []string{"eka", "dvi", "bahu"}

	for _, entry := range pronounMap {
		for i, form := range entry.forms {
			if form == word {
				results = append(results, map[string]any{
					"type":      "pronoun",
					"base_form": entry.base,
					"gender":    entry.gender,
					"case":      entry.cCase,
					"vacana":    vacanaList[i],
				})
			}
		}
	}

	return results
}